### PR TITLE
Release/r67 

### PR DESCRIPTION
**Pull request by jbeemster**

### CODE REVIEW

File: 2-collectors/scala-stream-collector/src/main/scala/com.snowplowanalytics.snowplow.collectors/package.scala [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+  /**
+   * Whether the sink is for good rows or bad rows
+   */
+  object InputType extends Enumeration {
+    type InputType = Value
+    val Good, Bad = Value
+  }
+
+  /**
+   * Case Class for holding both good and
```
Comment (alexanderdean) [inline 2-collectors/scala-stream-collector/src/main/scala/com.snowplowanalytics.snowplow.collectors/package.scala]: Case class small c

File: 2-collectors/scala-stream-collector/src/main/scala/com.snowplowanalytics.snowplow.collectors/package.scala [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+   * bad sinks for the Stream Collector.
+   *
+   * @param good
+   * @param bad
+   */
+  case class CollectorSinks(good: AbstractSink, bad: AbstractSink)
+
+  /**
+   * Case Class for holding the results of
```
Comment (alexanderdean) [inline 2-collectors/scala-stream-collector/src/main/scala/com.snowplowanalytics.snowplow.collectors/package.scala]: Case class small c

File: 2-collectors/scala-stream-collector/src/main/scala/com.snowplowanalytics.snowplow.collectors/scalastream/utils/SplitBatch.scala [hunk 3]
```diff
@@ -0,0 +1,224 @@
...
+   * Returns a Bad Row as a String
+   *
+   * @param size The size of the failed event
+   * @param errors A list of errors for this row
+   */
+  def getBadRow(size: Int, errors: List[String]): String = {
+    compact(
+      ("size" -> size) ~
+      ("errors" -> errors) ~
+      ("failure_tstamp" -> getTimestamp(System.currentTimeMillis()))
```
Comment (alexanderdean) [inline 2-collectors/scala-stream-collector/src/main/scala/com.snowplowanalytics.snowplow.collectors/scalastream/utils/SplitBatch.scala]: Nice

File: 2-collectors/scala-stream-collector/src/main/scala/com.snowplowanalytics.snowplow.collectors/scalastream/utils/SplitBatch.scala [hunk 4]
```diff
@@ -0,0 +1,224 @@
...
+  /**
+   * Returns an ISO valid timestamp
+   *
+   * @param tstamp The Timestamp to convert
+   * @return the formatted Timestamp
+   */
+  def getTimestamp(tstamp: Long): String = {
+    val dt = new DateTime(tstamp)
+    TstampFormat.print(dt)
+  }
```
Comment (alexanderdean) [inline 2-collectors/scala-stream-collector/src/main/scala/com.snowplowanalytics.snowplow.collectors/scalastream/utils/SplitBatch.scala]: Do these helpers ^^ need to be public?

File: 2-collectors/scala-stream-collector/src/test/scala/com.snowplowanalytics.snowplow.collectors.scalastream/PostSpec.scala [hunk 5]
```diff
@@ -29,7 +29,7 @@ import org.specs2.specification.{Scope,Fragments}
 import spray.testkit.Specs2RouteTest
 
 // Spray
-import spray.http.{DateTime,HttpHeader,HttpRequest,HttpCookie}
+import spray.http.{DateTime,HttpHeader,HttpRequest,HttpCookie,RemoteAddress}
```
Comment (alexanderdean) [inline 2-collectors/scala-stream-collector/src/test/scala/com.snowplowanalytics.snowplow.collectors.scalastream/PostSpec.scala]: spaces after commas

File: 2-collectors/scala-stream-collector/src/test/scala/com.snowplowanalytics.snowplow.collectors.scalastream/utils/SplitBatchSpec.scala [hunk 6]
```diff
@@ -0,0 +1,168 @@
...
+
+  "SplitBatch.split" should {
+
+    "Batch a list of strings based on size" in {
+
+      splitBatch.split(List("a", "b", "c"), 5, 1) must_==
+        SplitBatchResult(
+          List(List("c"),List("b", "a")),
+          Nil)
+
```
Comment (alexanderdean) [inline 2-collectors/scala-stream-collector/src/test/scala/com.snowplowanalytics.snowplow.collectors.scalastream/utils/SplitBatchSpec.scala]: Unnecessary whitespace

File: 2-collectors/scala-stream-collector/src/test/scala/com.snowplowanalytics.snowplow.collectors.scalastream/utils/SplitBatchSpec.scala [hunk 7]
```diff
@@ -0,0 +1,168 @@
...
+      val actual = SplitBatch.splitAndSerializePayload(payload, 100)
+
+      val errorJson = parse(new String(actual.bad.head))
+
+      errorJson \ "size" must_== JInt(1019)
+
+      errorJson \ "errors" must_== JArray(List(JString("Cannot split record with null body")))
+
+      actual.good must_== Nil
+    }
```
Comment (alexanderdean) [inline 2-collectors/scala-stream-collector/src/test/scala/com.snowplowanalytics.snowplow.collectors.scalastream/utils/SplitBatchSpec.scala]: Too much whitespace ^^

File: 2-collectors/scala-stream-collector/src/test/scala/com.snowplowanalytics.snowplow.collectors.scalastream/utils/SplitBatchSpec.scala [hunk 8]
```diff
@@ -0,0 +1,168 @@
...
+      val data = compact(render(
+        ("schema" -> "s") ~
+        ("data" -> List(
+          ("e" -> "se") ~ ("tv" -> "js-2.4.3"),
+          ("e" -> "se") ~ ("tv" -> "js-2.4.3")
+        ))))
+
+      payload.setBody(data)
+
+      payload.setPath("p" * 1000)
```
Comment (alexanderdean) [inline 2-collectors/scala-stream-collector/src/test/scala/com.snowplowanalytics.snowplow.collectors.scalastream/utils/SplitBatchSpec.scala]: Why all the whitespace?

File: 3-enrich/scala-common-enrich/src/test/scala/com.snowplowanalytics.snowplow.enrich.common/enrichments/registry/JavascriptScriptEnrichmentSpec.scala [hunk 9]
```diff
@@ -70,9 +74,10 @@ class JavascriptScriptEnrichmentSpec extends Specification with ValidationMatche
 
   def e1 = {
     val actual = JavascriptScriptEnrichment.compile("[")
-    val expected = "Error compiling JavaScript script: [javax.script.ScriptException: sun.org.mozilla.javascript.internal.EvaluatorException: syntax error (<Unknown Source>#5)]"
+    val expectedOracleJdk = Failure("Error compiling JavaScript script: [javax.script.ScriptException: sun.org.mozilla.javascript.internal.EvaluatorException: syntax error (<Unknown Source>#5)]")
+    val expectedOpenJdk = Failure("Error compiling JavaScript script: [javax.script.ScriptException: sun.org.mozilla.javascript.EvaluatorException: syntax error (<Unknown Source>#5)]")
 
-    actual must beFailing(expected)
+    List(expectedOracleJdk, expectedOpenJdk) must contain(actual)
```
Comment (alexanderdean) [inline 3-enrich/scala-common-enrich/src/test/scala/com.snowplowanalytics.snowplow.enrich.common/enrichments/registry/JavascriptScriptEnrichmentSpec.scala]: Nice

File: 3-enrich/scala-kinesis-enrich/src/main/scala/com.snowplowanalytics.snowplow.enrich.kinesis/SnowplowTracking.scala [hunk 10]
```diff
@@ -0,0 +1,163 @@
...
+
+    val appName = config.getString("app-id")
+
+    // Not yet used
+    val method = config.getString("method")
+
+    val emitter = AsyncEmitter.createAndStart(endpoint, port)
+
+    new Tracker(List(emitter), generated.Settings.name, appName)
+  }
```
Comment (alexanderdean) [inline 3-enrich/scala-kinesis-enrich/src/main/scala/com.snowplowanalytics.snowplow.enrich.kinesis/SnowplowTracking.scala]: Too much whitespace ^^

File: 3-enrich/scala-kinesis-enrich/src/main/scala/com.snowplowanalytics.snowplow.enrich.kinesis/sinks/KinesisSink.scala [hunk 11]
```diff
@@ -171,7 +174,7 @@ class KinesisSink(provider: AWSCredentialsProvider,
     def addEvent(event: (ByteBuffer, String)) {
       val newBytes = event._1.capacity
 
-      if (newBytes >= 51200) {
+      if (newBytes >= 1000000L) {
```
Comment (alexanderdean) [inline 3-enrich/scala-kinesis-enrich/src/main/scala/com.snowplowanalytics.snowplow.enrich.kinesis/sinks/KinesisSink.scala]: Move magic constant up into top of class

File: 3-enrich/scala-kinesis-enrich/src/main/scala/com.snowplowanalytics.snowplow.enrich.kinesis/sources/AbstractSource.scala [hunk 12]
```diff
@@ -63,14 +63,71 @@ import common.ValidatedMaybeCollectorPayload
...
  * we support.
  */
 abstract class AbstractSource(config: KinesisEnrichConfig, igluResolver: Resolver,
-                              enrichmentRegistry: EnrichmentRegistry) {
+                              enrichmentRegistry: EnrichmentRegistry, 
+                              tracker: Option[Tracker]) {
   
-  val MaxRecordSize = 51200
+  val MaxRecordSize = if (config.sink == Sink.Kinesis) {
+    Some(1000000L)
```
Comment (alexanderdean) [inline 3-enrich/scala-kinesis-enrich/src/main/scala/com.snowplowanalytics.snowplow.enrich.kinesis/sources/AbstractSource.scala]: Duplication of magic constant from previous comment above - centralize somwhere

File: 3-enrich/scala-kinesis-enrich/src/main/scala/com.snowplowanalytics.snowplow.enrich.kinesis/sources/KinesisSource.scala [hunk 13]
```diff
@@ -97,7 +100,7 @@ class KinesisSource(config: KinesisEnrichConfig, igluResolver: Resolver, enrichm
 
     val rawEventProcessorFactory = new RawEventProcessorFactory(
       config,
-      sink.get // TODO: yech
+      sink.get.get // TODO: yech
```
Comment (alexanderdean) [inline 3-enrich/scala-kinesis-enrich/src/main/scala/com.snowplowanalytics.snowplow.enrich.kinesis/sources/KinesisSource.scala]: TODO: yech, yech

File: 3-enrich/scala-kinesis-enrich/src/test/scala/com.snowplowanalytics.snowplow.enrich.kinesis/sources/AbstractSourceSpec.scala [hunk 14]
```diff
@@ -0,0 +1,71 @@
...
+    "get the size of a string of ASCII characters" in {
+
+      AbstractSource.getSize("abcdefg") must_== 7
+    }
+
+    "get the size of a string containing non-ASCII characters" in {
+
+      AbstractSource.getSize("™®字") must_== 8
+    }
+  }
```
Comment (alexanderdean) [inline 3-enrich/scala-kinesis-enrich/src/test/scala/com.snowplowanalytics.snowplow.enrich.kinesis/sources/AbstractSourceSpec.scala]: Too much whitespace above

File: 4-storage/kinesis-elasticsearch-sink/src/main/scala/com.snowplowanalytics.snowplow.storage.kinesis/elasticsearch/ElasticsearchPipeline.scala [hunk 15]
```diff
@@ -38,17 +39,30 @@ import com.amazonaws.services.kinesis.connectors.impl.{BasicMemoryBuffer,AllPass
...
 
+// Tracker
+import scalatracker.Tracker
+
 /**
 * ElasticsearchPipeline class sets up the Emitter/Buffer/Transformer/Filter
 */
-class ElasticsearchPipeline(streamType: StreamType, documentIndex: String, documentType: String, goodSink: Option[ISink], badSink: ISink)
-  extends IKinesisConnectorPipeline[ValidatedRecord, EmitterInput] {
+class ElasticsearchPipeline(
```
Comment (alexanderdean) [inline 4-storage/kinesis-elasticsearch-sink/src/main/scala/com.snowplowanalytics.snowplow.storage.kinesis/elasticsearch/ElasticsearchPipeline.scala]: Put a docstring on with a tparam for each field

File: 4-storage/kinesis-elasticsearch-sink/src/main/scala/com.snowplowanalytics.snowplow.storage.kinesis/elasticsearch/SnowplowElasticsearchEmitter.scala [hunk 16]
```diff
@@ -286,7 +317,11 @@ class SnowplowElasticsearchEmitter(configuration: KinesisConnectorConfiguration,
   override def fail(records: JList[EmitterInput]) {
     records foreach {
       record => {
-        val output = compact(render(("line" -> record._1) ~ ("errors" -> record._2.swap.getOrElse(Nil))))
+        val output = compact(render(
+          ("line" -> record._1) ~ 
+          ("errors" -> record._2.swap.getOrElse(Nil)) ~
+          ("failure_tstamp" -> getTimestamp(System.currentTimeMillis()))
```
Comment (alexanderdean) [inline 4-storage/kinesis-elasticsearch-sink/src/main/scala/com.snowplowanalytics.snowplow.storage.kinesis/elasticsearch/SnowplowElasticsearchEmitter.scala]: Good

File: 4-storage/kinesis-elasticsearch-sink/src/main/scala/com.snowplowanalytics.snowplow.storage.kinesis/elasticsearch/SnowplowElasticsearchEmitter.scala [hunk 17]
```diff
@@ -317,6 +352,24 @@ class SnowplowElasticsearchEmitter(configuration: KinesisConnectorConfiguration,
   }
 
   /**
+   * Terminate the application in a way the KCL cannot stop
+   *
+   * Prevents shutdown hooks from running
+   */
+  private def forcibleShutdown() {
```
Comment (alexanderdean) [inline 4-storage/kinesis-elasticsearch-sink/src/main/scala/com.snowplowanalytics.snowplow.storage.kinesis/elasticsearch/SnowplowElasticsearchEmitter.scala]: forcible -> force for a more verby method

File: 4-storage/kinesis-elasticsearch-sink/src/main/scala/com.snowplowanalytics.snowplow.storage.kinesis/elasticsearch/SnowplowTracking.scala [hunk 18]
```diff
@@ -0,0 +1,151 @@
...
+import com.snowplowanalytics.snowplow.scalatracker.Tracker
+import com.snowplowanalytics.snowplow.scalatracker.SelfDescribingJson
+import com.snowplowanalytics.snowplow.scalatracker.emitters.AsyncEmitter
+
+/**
+ * Functionality for sending Snowplow events for monitoring purposes
+ */
+object SnowplowTracking {
+
+  val HeartbeatInterval = 300000L
```
Comment (alexanderdean) [inline 4-storage/kinesis-elasticsearch-sink/src/main/scala/com.snowplowanalytics.snowplow.storage.kinesis/elasticsearch/SnowplowTracking.scala]: Public?

File: 4-storage/kinesis-elasticsearch-sink/src/main/scala/com.snowplowanalytics.snowplow.storage.kinesis/elasticsearch/SnowplowTracking.scala [hunk 19]
```diff
@@ -0,0 +1,151 @@
...
+
+    val appName = config.getString("app-id")
+
+    // Not yet used
+    val method = config.getString("method")
+
+    val emitter = AsyncEmitter.createAndStart(endpoint, port)
+
+    new Tracker(List(emitter), generated.Settings.name, appName)
+  }
```
Comment (alexanderdean) [inline 4-storage/kinesis-elasticsearch-sink/src/main/scala/com.snowplowanalytics.snowplow.storage.kinesis/elasticsearch/SnowplowTracking.scala]: Too much whitespace

File: 4-storage/kinesis-elasticsearch-sink/src/main/scala/com.snowplowanalytics.snowplow.storage.kinesis/elasticsearch/SnowplowTracking.scala [hunk 20]
```diff
@@ -0,0 +1,151 @@
...
+      ("failureCount" -> failureCount) ~
+      ("initialFailureTime" -> initialFailureTime) ~
+      ("message" -> message)
+    ))
+  }
+
+  /**
+   * Send an initialization event and schedule heartbeat and shutdown events
+   *
+   * @param tracker
```
Comment (alexanderdean) [inline 4-storage/kinesis-elasticsearch-sink/src/main/scala/com.snowplowanalytics.snowplow.storage.kinesis/elasticsearch/SnowplowTracking.scala]: Add comment for tracker arg

File: 4-storage/kinesis-elasticsearch-sink/src/main/scala/com.snowplowanalytics.snowplow.storage.kinesis/elasticsearch/SnowplowTracking.scala [hunk 21]
```diff
@@ -0,0 +1,151 @@
...
+      }
+    }
+
+    heartbeatThread.start()
+  }
+
+  /**
+   * Send an application_initialized unstructured event
+   *
+   * @param tracker
```
Comment (alexanderdean) [inline 4-storage/kinesis-elasticsearch-sink/src/main/scala/com.snowplowanalytics.snowplow.storage.kinesis/elasticsearch/SnowplowTracking.scala]: Add comment for tracker arg

File: 4-storage/kinesis-elasticsearch-sink/src/main/scala/com.snowplowanalytics.snowplow.storage.kinesis/elasticsearch/SnowplowTracking.scala [hunk 22]
```diff
@@ -0,0 +1,151 @@
...
+    tracker.trackUnstructEvent(SelfDescribingJson(
+      "iglu:com.snowplowanalytics.snowplow/application_initialized/jsonschema/1-0-0",
+      JObject(Nil)
+    ))
+  }
+
+  /**
+   * Send an application_shutdown unstructured event
+   *
+   * @param tracker
```
Comment (alexanderdean) [inline 4-storage/kinesis-elasticsearch-sink/src/main/scala/com.snowplowanalytics.snowplow.storage.kinesis/elasticsearch/SnowplowTracking.scala]: Add comment for tracker arg

File: 4-storage/redshift-storage/jsonpaths/com.snowplowanalytics.monitoring/sink_write_failed_1.json [hunk 23]
```diff
@@ -0,0 +1,21 @@
...
+    "$.hierarchy.refRoot",
+    "$.hierarchy.refTree",
+    "$.hierarchy.refParent",
+
+    "$.data.sink",
+    "$.data.failureCount",
+    "$.data.initialFailureTime",
+    "$.data.lastRetryPeriod",
+    "$.data.message"
+    ]
```
Comment (alexanderdean) [inline 4-storage/redshift-storage/jsonpaths/com.snowplowanalytics.monitoring/sink_write_failed_1.json]: Nice

File: 3-enrich/scala-kinesis-enrich/src/main/scala/com.snowplowanalytics.snowplow.enrich.kinesis/sources/AbstractSource.scala (hunk 12 above)
Comment (jbeemster) [inline 3-enrich/scala-kinesis-enrich/src/main/scala/com.snowplowanalytics.snowplow.enrich.kinesis/sources/AbstractSource.scala]: Should I make a package for this app to put the magic constant in?

Comment (alexanderdean) [inline 3-enrich/scala-kinesis-enrich/src/main/scala/com.snowplowanalytics.snowplow.enrich.kinesis/sources/AbstractSource.scala]: Sounds good to me...
